apstools/utils.py

```diff
@@ -384,7 +393,377 @@ def itemizer(fmt, items):
     return [fmt % k for k in items]
 
 
+def getCatalog(ref=None):
+    if isinstance(ref, str):  # and ref in databroker.catalog:
+        return databroker.catalog[ref]
+    if ref is not None and hasattr(ref, "v2"):
+        return ref.v2
+
+    cat = getDefaultCatalog()
+    if cat is None:
+        raise ValueError("Cannot identify default databroker catalog.")
+    return cat
+
+
+def findCatalogsInNamespace():
+    g = {}
+    g.update(getDefaultNamespace())
+    ns_cats = {}
+    for k, v in g.items():
+        if not k.startswith("_") and hasattr(v, "__class__"):
+            try:
+                if (hasattr(v.v2, "container") and hasattr(v.v2, "metadata")):
+                    ns_cats[k] = v
+            except (AttributeError, TypeError):
+                continue
+    return ns_cats
+
+
+def getDefaultCatalog():
+    cats = findCatalogsInNamespace()
+    if len(cats) == 1:
+        return cats[list(cats.keys())[0]]
+    if len(cats) > 1:
+        choices = '   '.join([
+            f"{k} ({v.name})"
+            for k, v in cats.items()
+        ])
+        raise ValueError(
+            "No catalog defined.  "
+            "Multiple catalog objects available."
+            "  Specify one of these:"
+            f" {choices}"
+        )
+
+    cats = list(databroker.catalog)
+    if len(cats) == 1:
+        return databroker.catalog[cats[0]]
+    if len(cats) > 1:
+        choices = '   '.join([f'databroker.catalog[\"{k}\"]' for k in cats])
+        raise ValueError(
+            "No catalog defined.  "
+            "Multiple catalog configurations available."
+            "  Create specific catalog object from one of these commands:"
+            f" {choices}"
+        )
+
+    raise ValueError("No catalogs available.")
+
+
+def getDefaultNamespace():
+    """
+    get the IPython shell's namespace dictionary (or globals() if not found)
+    """
+    try:
+        from IPython import get_ipython
+
+        ns = get_ipython().user_ns
+    except (ModuleNotFoundError, AttributeError):
+        ns = globals()
+    return ns
+
+
+@dataclass
+class ListRuns:
+    """
+    List the runs from the given catalog according to some options.
+
+    EXAMPLE::
+
+        ListRuns(cat).to_dataframe()
+
+    PUBLIC METHODS
+
+    .. autosummary::
+
+        ~to_dataframe
+        ~to_table
+        ~parse_runs
+
+    INTERNAL METHODS
+
+    .. autosummary::
+
+        ~_get_by_key
+        ~_check_cat
+        ~_apply_search_filters
+        ~_sorter
+        ~_check_keys
+
+    """
+
+    cat: object = None
+    query: dict = None
+    keys: str = None
+    missing: str = ""
+    num: int = 20
+    reverse: bool = True
+    since: str = None
+    sortby: str = "time"
+    timefmt: str = "%Y-%m-%d %H:%M:%S"
+    until: str = None
+
+    _default_keys = "scan_id time plan_name detectors"
+
+    def _get_by_key(self, md, key):
```

Comment:
Should this always search both "start" and "stop" docs (reporting missing if not there)? In one hand, it is simpler to handle it as is, and the user can force searching the "stop" doc anyway. But in the other hand, this assumes that for most metadata, only one of these docs will matter. That is true today, with only "time" as an exception.

Reply:
Considering how to lower the barrier to use for those who do not read documentation.